### PR TITLE
[Speculative Decoding] Enabling bonus token in speculative decoding for KV cache based models

### LINKED ISSUE #4212: [Speculative decoding] [Performance]: Re-enable bonus tokens
### Proposal to improve performance

In https://github.com/vllm-project/vllm/pull/3951 we disable bonus tokens (token sampled from verifier model assuming all proposal tokens are accepted) because its KV is not generated for the draft model. We can fix this by "prefilling" the KV of bonus tokens in the draft model. Note that for proposal methods not requiring KV (e.g. prompt lookup), we can re-enable bonus tokens and get a speedup there.

The impact of this performance improvement depends on the speculation length. For low K, e.g. 1, where the probability of accepting the single spec token is high (~= how aligned the draft model and target model are on the sequence), it has high impact because accepting 1 token allows us to emit 2 tokens (1 speculative and 1 bonus). Since we disable bonus tokens, we can now only emit 1 token (the accepted speculative one).

For higher K the impact is less as the likelihood of accepting all speculative tokens is exponentially lower.

https://github.com/vllm-project/vllm/blob/323f27b9048713cdbab31995265975842a937167/vllm/model_executor/layers/rejection_sampler.py#L311-L315

### PR DESCRIPTION
**Pull request by sroy745**
FILL IN THE PR DESCRIPTION HERE

FIX #4212 

In this PR we make the following changes

1. Update the spec_decode_worker to keep track of the sequence_ids which we were assigned bonus token ids in their last forward pass. We record this only for the MultiStepWorker since other Worker types don't utilize the KV cache for token generation.  **Currently we don't clear out the sequence ids from this list even on sequence termination. We need a way to get notified on sequence termination and remove those sequence ids** 
2. Updated the MultiStepWorker to expand the batch during step 0. During batch expansion we check to see which of the sequence ids were assigned bonus tokens in their last forward pass. For each of those sequences we add a new sequence (with the same seq_id) without the bonus token. Once the forward pass for step 0 is completed we filter out the response to retain only those responses which correspond to the original sequences.
3. Added a flag --disable-bonus-tokens-in-kv-cache to enable/disable bonus tokens for MultiStepWorker.

Some numbers from e2e tests. Note that the e2e tests don't use cuda graphs. The draft model is JackFram/llama-68m and the target model is JackFram/llama-160m. We use a batch size of 64. Completion time for num_speculation = 1 shows ~33% speedup

- w/o bonus Processed prompts: 100%|█████████████████64/64 [00:06<00:00, 10.13it/s, est. speed input: 78.48 toks/s, output: 2592.40 toks/s]

- with bonus Processed prompts: 100%|█████████████████████ 64/64 [00:04<00:00, 15.50it/s, est. speed input: 120.13 toks/s, output: 3968.28 toks/s]

### CODE REVIEW

**Comment by cadedaniel** [comment]
LMK once it's ready for review @sroy745

File: vllm/spec_decode/ngram_worker.py [hunk 1]
```diff
@@ -133,12 +136,15 @@ def sampler_output(
     def get_spec_proposals(
         self,
         execute_model_req: ExecuteModelRequest,
+        # Unused parameter. NGramWorker does not use the KV Cache and
+        # therefore does not need this parameter.
+        seq_ids_with_bonus_token_in_last_step: set = set(),
```
Comment (cadedaniel) [inline vllm/spec_decode/ngram_worker.py:141]: btw this is bad python antipattern; the object will be re-used over the lifetime of the function, see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: tests/spec_decode/e2e/test_multistep_correctness.py [hunk 2]
```diff
@@ -611,3 +611,54 @@ def test_many_k(baseline_llm_generator, test_llm_generator, batch_size: int,
...
+    """
+    Test speculative decoding with bonus token acceptance enabled for a model 
+    using KV cache.
+
+    This test verifies that speculative decoding produces results exactly 
+    matching those from non-speculative decoding across various values of 'k' 
+    and batch sizes when bonus token acceptance is enabled. It ensures 
+    correctness by comparing the output of speculative decoding with the baseline.
+    """
+    run_greedy_equality_correctness_test(baseline_llm_generator,
```
Comment (cadedaniel) [inline tests/spec_decode/e2e/test_multistep_correctness.py:660]: Test looks good. Can you manually check the draft acceptance rate? for the same draft/target model it should be 100%. without your fix (and with bonus token enabled) the acceptance rate goes to like 80%.

Ideally we have a test for this, not sure how easy that is..

File: vllm/engine/arg_utils.py [hunk 3]
```diff
@@ -567,6 +568,15 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
             help='Min size of window for ngram prompt lookup in speculative '
             'decoding.')
 
+        parser.add_argument(
```
Comment (cadedaniel) [inline vllm/engine/arg_utils.py:571]: I feel we can remove this flag completely in spec decode if your change fixes the bug.

File: vllm/spec_decode/multi_step_worker.py [hunk 4]
```diff
@@ -85,15 +99,116 @@ def sampler_output(
...
+        Returns:
+            Tuple[ExecuteModelRequest, List[int]]: The updated execute model request with expanded 
+            sequences and a list of indices corresponding to the original sequence groups.
+        """
+        updated_seq_group_metadata_list = []
+        updated_execute_model_req = execute_model_req.clone(updated_seq_group_metadata_list)
+        indices_of_original_sequence_groups = []
+        for seq_group in execute_model_req.seq_group_metadata_list:
+            seq_ids_with_bonus_tokens = []
+            for seq_id, seq_data in seq_group.seq_data.items():
```
Comment (cadedaniel) [inline vllm/spec_decode/multi_step_worker.py:128]: approach looks good. pretty messy but not anything easy to fix that.

File: tests/spec_decode/test_spec_decode_worker.py [hunk 5]
```diff
@@ -527,6 +530,165 @@ def test_empty_input_batch(k: int, batch_size: int):
     draft_worker.execute_model.assert_called_once_with(execute_model_req)
     target_worker.execute_model.assert_called_once_with(execute_model_req)
 
+@torch.inference_mode()
+def test_populate_seq_ids_with_bonus_tokens():
```
Comment (cadedaniel) [inline tests/spec_decode/test_spec_decode_worker.py:534]: one edge case is where a sequence is skipped but is present in `seq_with_bonus_token_in_last_step`

**Review by cadedaniel — COMMENTED** [review]
approach looks good. seems this will conflict with https://github.com/vllm-project/vllm/pull/5799. we should work out a way to combine both.

**Comment by sroy745** [comment]
Thanks for the review. I will continue to address the comments. However I will probably wait for pr (https://github.com/vllm-project/vllm/pull/5799) to be merged before rebasing and pinging for the second round of reviews.

File: vllm/engine/arg_utils.py (hunk 3 above)
Comment (sroy745) [reply vllm/engine/arg_utils.py]: Done

File: tests/spec_decode/test_spec_decode_worker.py (hunk 5 above)
Comment (sroy745) [reply tests/spec_decode/test_spec_decode_worker.py:656]: Added this case in Forward pass # 2.

File: vllm/spec_decode/multi_step_worker.py (hunk 4 above)
Comment (sroy745) [reply vllm/spec_decode/multi_step_worker.py]: Tried to simplify it a bit by moving some of the logic to 2 helper functions  _shallow_copy_seq_group_metadata and _copy_seq_metadata_excluding_last_token. PTAL.

File: tests/spec_decode/e2e/test_multistep_correctness.py (hunk 2 above)
Comment (sroy745) [reply tests/spec_decode/e2e/test_multistep_correctness.py]: Yeah I thought of adding such an e2e test but I could find an easy way to access the metrics_collector and the stats. As you suggested I added an e2e test in test_multistep_correctness.py with both draft and target model as same ("JackFram/llama-68m") and number of speculation as 3. The System efficiency with bonus token enabled is 1 and with bonus token disabled is 0.75

**With bonus token**
INFO 07-08 06:27:27 metrics.py:316] Speculative metrics: Draft acceptance rate: 1.000, System efficiency: 1.000, Number of speculative tokens: 3, Number of accepted tokens: 21696, Number of draft tokens tokens: 21696, Number of emitted tokens tokens: 28928.

**Without bonus token**
Speculative metrics: Draft acceptance rate: 1.000, System efficiency: 0.750, Number of speculative tokens: 3, Number of accepted tokens: 21696, Number of draft tokens tokens: 21696, Number of emitted tokens tokens: 21696.

File: vllm/spec_decode/ngram_worker.py (hunk 1 above)
Comment (sroy745) [reply vllm/spec_decode/ngram_worker.py]: Done

**Comment by cadedaniel** [comment]
Awesome. Will take a look tomorrow.

File: vllm/spec_decode/multi_step_worker.py [hunk 6]
```diff
@@ -123,36 +222,66 @@ def _append_new_tokens(
...
+        # Shallow-copy the SequenceGroupMetadata.
+        new_seq_group_metadata = copy.copy(seq_group_metadata)
+        # Shallow-copy seq_data and modify the output_token_ids.
+        new_seq_data: Dict[int, SequenceData] = {}
+        for seq_id, old_seq_data in seq_group_metadata.seq_data.items():
+            if (seq_id in seq_ids_to_copy):
+                new_seq_data[seq_id] = copy.copy(old_seq_data)
+                # Copy all the output token ids except the last.
+                # Also reduce num_computed_tokens by 1 since we are not
+                # including the last output token.
```
Comment (cadedaniel) [inline vllm/spec_decode/multi_step_worker.py:279]: Can we add more comment here motivating this? I thought the worker's themselves don't take this into account unless chunked prefill is enabled.

File: vllm/spec_decode/spec_decode_worker.py [hunk 7]
```diff
@@ -127,7 +125,6 @@ def create_worker(
 
             if draft_worker_kwargs[
                     "model_config"].hf_config.model_type == "mlp_speculator":
-                disable_bonus_tokens = False
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:130]: can we either remove the `disable_bonus_token` paths entirely, or log below whether the bonus token is enabled?

File: vllm/spec_decode/spec_decode_worker.py [hunk 8]
```diff
@@ -195,6 +192,19 @@ def __init__(
...
+        # used for token generation such as in the case of MultiStepWorker.
+        self.seq_with_bonus_token_in_last_step: Set[int] = set()
+        # Tracks the currently active request ids and the sequence IDs
+        # corresponding to them
+        self.request_id_seq_id_mapping: Dict[str, Set[int]] = defaultdict(set)
+        # Tracks if the proposer worker uses the KV cache or not.
+        self.proposer_uses_kv_cache = True
+        if (isinstance(self.proposer_worker,
+                       (MLPSpeculatorWorker, NGramWorker))):
+            self.proposer_uses_kv_cache = False
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:206]: let's move this logic to `create_worker`? we want to minimize data-dependent control flow in constructor much as possible so testing is easier

File: vllm/spec_decode/spec_decode_worker.py [hunk 9]
```diff
@@ -195,6 +192,19 @@ def __init__(
         self._metrics = AsyncMetricsCollector(
             self.spec_decode_sampler
         ) if metrics_collector is None else metrics_collector
+        # Tracks the sequence IDs that received a bonus token ID in
+        # their last forward pass. Needed only if KV cache is being
+        # used for token generation such as in the case of MultiStepWorker.
+        self.seq_with_bonus_token_in_last_step: Set[int] = set()
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:198]: nit: make these fields private `_ seq_with_bonus_token_in_last_step`

File: vllm/spec_decode/spec_decode_worker.py [hunk 10]
```diff
@@ -302,6 +312,12 @@ def execute_model(
             broadcast_tensor_dict({}, src=0)
             return []
 
+        if self.proposer_uses_kv_cache:
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:315]: nit: preference for always running this code so the state is the same with and without `proposer_uses_kv_caches`. this will reduce branches in code and number of states we have to reason about.

Comment (cadedaniel) [reply vllm/spec_decode/spec_decode_worker.py:315]: since we have small batch sizes, I expect latency hit to be <50µs

File: vllm/spec_decode/spec_decode_worker.py [hunk 11]
```diff
@@ -612,22 +631,31 @@ def _create_output_sampler_list(
...
+            # The proposer worker uses KV cache. Hence populate the data
+            # structures needed to keep track of bonus tokens.
+            for seq_index, seq_id in enumerate(seq_ids):
+                last_token_id = accepted_token_ids_by_step[-1][seq_index]
+                if last_token_id == -1:
+                    self.seq_with_bonus_token_in_last_step.discard(seq_id)
+                else:
+                    self.seq_with_bonus_token_in_last_step.add(seq_id)
+            for request_id, seq_ids in request_ids_seq_ids_mapping.items():
+                self.request_id_seq_id_mapping[request_id].update(seq_ids)
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:652]: nit: suggest moving to own function so someone reading this function doesn't have to know about all these details

File: vllm/spec_decode/spec_decode_worker.py [hunk 12]
```diff
@@ -603,7 +622,7 @@ def _create_output_sampler_list(
             for sequence_index in range(batch_size):
                 # Each sequence may have a different num_logprobs; retrieve it.
                 num_logprobs = num_logprobs_per_seq[sequence_index]
-
+                seq_id = seq_ids[sequence_index]
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:625]: nit: remove change

File: vllm/spec_decode/spec_decode_worker.py (hunk 10 above)
Comment (cadedaniel) [reply vllm/spec_decode/spec_decode_worker.py:315]: also suggest moving to own function so reader of this method doesn't have to know these details

File: tests/spec_decode/test_multi_step_worker.py [hunk 13]
```diff
@@ -277,6 +280,124 @@ def test_same_output_for_multi_step():
...
+    )
+
+    worker = create_worker(
+        Worker,
+        model_name,
+        block_size,
+        num_gpu_blocks,
+        seed,
+    )
+    # Make sure we go over the block boundary.
```
Comment (cadedaniel) [inline tests/spec_decode/test_multi_step_worker.py:314]: comment seems to be floating // not applicable to anything

File: tests/spec_decode/test_multi_step_worker.py [hunk 14]
```diff
@@ -277,6 +280,124 @@ def test_same_output_for_multi_step():
...
+        elif (continuations[index][-1] != output.samples[0].output_token):
+            num_mismatch += 1
+    # The prediction is accurate for some of the sequences even without proper
+    # handling of the bonus tokens. Hence verify that the number of sequences
+    # for which there is a mismatch is > 0.
+    assert (num_mismatch > 0)
+
+    # Make a second run. In this run correctly specify that all the sequences
+    # have bonus tokens. With this setup verify that the third token
+    # prediction is accurate for all sequences.
```
Comment (cadedaniel) [inline tests/spec_decode/test_multi_step_worker.py:389]: This test is pretty big -- can we separate out into the happy path (that it is correct when `seq_ids_with_bonus_token_in_last_step` is complete) and the sad path (that it is incorrect when `seq_ids_with_bonus_token_in_last_step` is incomplete)?

File: tests/spec_decode/test_multi_step_worker.py [hunk 15]
```diff
@@ -440,3 +564,116 @@ def test_draft_proposals_mixed_k():
     assert proposals.proposal_lens.tolist() == [
         k for _ in range(expected_num_proposal_seqs - 1)
     ] + [0 for _ in range(expected_num_no_proposal_seqs)] + [k]
+
+
+@torch.inference_mode()
+def test_expand_execute_model_request_for_bonus_tokens():
```
Comment (cadedaniel) [inline tests/spec_decode/test_multi_step_worker.py:570]: Why do we need both this test and `test_same_output_for_multi_step_with_batch_expansion`?

Comment (cadedaniel) [reply tests/spec_decode/test_multi_step_worker.py:570]: fine to have both but we should add something in the docstring which explains the difference

File: tests/spec_decode/test_multi_step_worker.py [hunk 16]
```diff
@@ -440,3 +564,116 @@ def test_draft_proposals_mixed_k():
...
+    target_token_logprobs = torch.rand(batch_size,
+                                       num_steps,
+                                       vocab_size,
+                                       dtype=torch.float32,
+                                       device='cuda')
+    sampler_output_list = create_sampler_output_list(target_token_ids,
+                                                     target_token_probs,
+                                                     target_token_logprobs)
+    output_indices_to_retain = random.sample(range(num_steps),
+                                             max(1, num_steps // 2))
```
Comment (cadedaniel) [inline tests/spec_decode/test_multi_step_worker.py:671]: I am confused why we sample from `range(num_steps)`. shouldn't it be `range(batch_size)`?

File: tests/spec_decode/test_spec_decode_worker.py [hunk 17]
```diff
@@ -645,3 +650,249 @@ def test_split_num_cache_blocks_evenly(available_gpu_blocks: int,
...
+        range(batch_size), num_sequences_with_bonus_tokens)
+    # Create a mask that is True for indices in seq_indexes_with_bonus_tokens
+    mask = torch.ones(batch_size, dtype=torch.bool, device='cuda')
+    mask[seq_indexes_with_bonus_tokens] = False
+    # Forward Pass : 0
+    # Set the last token ID to -1 for all indices not in
+    # seq_indexes_with_bonus_tokens to indicate the lack of bonus token in
+    # those indices.
+    accepted_token_ids[mask, -1:] = -1
+    worker = SpecDecodeWorker(draft_worker,
```
Comment (cadedaniel) [inline tests/spec_decode/test_spec_decode_worker.py:711]: Can we break the three tests in this test into their own tests? it will be easier to follow and debug

**Review by cadedaniel — COMMENTED** [review]
Can we manually verify the following?
* When bonus tokens are enabled and the same model is used for draft and target, we get 100% draft acceptance rate. This indicates that the KV of the draft model is ~equal to the KV of the target model.

**Comment by cadedaniel** [comment]
Oh I just saw your response

> Yeah I thought of adding such an e2e test but I could find an easy way to access the metrics_collector and the stats. As you suggested I added an e2e test in test_multistep_correctness.py with both draft and target model as same ("JackFram/llama-68m") and number of speculation as 3. The System efficiency with bonus token enabled is 1 and with bonus token disabled is 0.75
>
> With bonus token
INFO 07-08 06:27:27 metrics.py:316] Speculative metrics: Draft acceptance rate: 1.000, System efficiency: 1.000, Number of speculative tokens: 3, Number of accepted tokens: 21696, Number of draft tokens tokens: 21696, Number of emitted tokens tokens: 28928.
> 
> Without bonus token
Speculative metrics: Draft acceptance rate: 1.000, System efficiency: 0.750, Number of speculative tokens: 3, Number of accepted tokens: 21696, Number of draft tokens tokens: 21696, Number of emitted tokens tokens: 21696.

This is awesome! so exciting to see this working!

File: vllm/spec_decode/spec_decode_worker.py (hunk 7 above)
Comment (sroy745) [reply vllm/spec_decode/spec_decode_worker.py:130]: Removed disable_bonus_token.

File: vllm/spec_decode/spec_decode_worker.py (hunk 8 above)
Comment (sroy745) [reply vllm/spec_decode/spec_decode_worker.py]: I removed this completely. We are always maintaining the bonus token related data structures even if it might not be needed by the worker.

File: vllm/spec_decode/spec_decode_worker.py (hunk 10 above)
Comment (sroy745) [reply vllm/spec_decode/spec_decode_worker.py]: Done. Always running this logic. Also moved it to its own function.

File: tests/spec_decode/test_multi_step_worker.py (hunk 14 above)
Comment (sroy745) [reply tests/spec_decode/test_multi_step_worker.py]: I split it up into 2 tests.

File: tests/spec_decode/test_multi_step_worker.py (hunk 15 above)
Comment (sroy745) [reply tests/spec_decode/test_multi_step_worker.py]: Removed this test as well as test_filter_model_output. Both should be covered by test_same_output_for_multi_step_with_batch_expansion as you suggested.

File: tests/spec_decode/test_multi_step_worker.py (hunk 16 above)
Comment (sroy745) [reply tests/spec_decode/test_multi_step_worker.py]: It should be range(batch_size). I removed the test though in favor of test_same_output_for_multi_step_with_batch_expansion

File: tests/spec_decode/test_spec_decode_worker.py (hunk 17 above)
Comment (sroy745) [reply tests/spec_decode/test_spec_decode_worker.py]: I modified the test to initialize the internal data structures with some fake data and then run a forward pass to simulate all 3 cases. PTAL. It should be simpler to follow now while covering all the cases.

File: tests/spec_decode/test_multi_step_worker.py (hunk 13 above)
Comment (sroy745) [reply tests/spec_decode/test_multi_step_worker.py]: Removed.

File: vllm/spec_decode/spec_decode_worker.py (hunk 9 above)
Comment (sroy745) [reply vllm/spec_decode/spec_decode_worker.py]: Done

File: vllm/spec_decode/multi_step_worker.py (hunk 6 above)
Comment (sroy745) [reply vllm/spec_decode/multi_step_worker.py:279]: I included this to keep the value consistent. It probably not needed since it gets used only for chunked prefill. However I see this being updated elsewhere in the sd code e.g. https://sourcegraph.com/github.com/vllm-project/vllm/-/blob/vllm/spec_decode/draft_model_runner.py?L116 . Hence I added this. Is this confusing? Should I remove it?

**Comment by sroy745** [comment]
Thanks for the review. Addressed the comments. PTAL

File: vllm/spec_decode/multi_step_worker.py (hunk 6 above)
Comment (cadedaniel) [reply vllm/spec_decode/multi_step_worker.py:279]: Gotcha -- yeah can leave in. Add a comment?

**Review by cadedaniel — APPROVED** [review]
Looks great!

**Review by sroy745 — COMMENTED** [review]
Thanks for the review. Addressed your comment and rebased. Should be ready to merge once the tests complete.

**Comment by cadedaniel** [comment]
Merged!
